src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -68,33 +66,30 @@ public boolean authenticate(String username, PublicKey key, ServerSession sessio
         return u;
     }
 
-    private @CheckForNull Authentication verifyUserUsingSecurityRealm(@Nonnull User user) {
+    private @CheckForNull UserDetails validateUserDetails(@Nonnull User user) {
         try {
-            return user.impersonate();
+            UserDetails userDetails = user.getUserDetailsForImpersonation();
+            if (!userDetails.isEnabled()) {
+                LOGGER.fine(() -> user.getId() + " account is disabled");
+                return null;
+            } else if (!userDetails.isAccountNonLocked()) {
+                LOGGER.fine(() -> user.getId() + " account is locked");
+                return null;
+            } else if (!userDetails.isAccountNonExpired()) {
+                LOGGER.fine(() -> user.getId() + " account is expired");
+                return null;
+            } else if (!userDetails.isCredentialsNonExpired()) {
+                LOGGER.fine(() -> user.getId() + " account credentials are expired");
+                return null;
```

Comment:
I thought we were going to implement these checks in the `SecurityRealm` where meaningful? Or is this just a second line of defense?

Reply:
Little bit of both. This helps explore whether or not it's possible to implement this feature without patching core as it's less intrusive. The core patch would also limit _any_ calls to `impersonate()` on that user, while this approach is working to only limit alternative authentication methods. There may still be a patch necessary for core, though. I'm going to look more into that.

Reply:
Ok, upon further inspection, I believe we can still patch core. This will be as additional hardening (plus makes it easier to test the changes).

Reply:
Updated https://github.com/jenkinsci/jenkins/pull/4925